File: rust/src/repo.rs
use std::fs;
use std::io;
use std::path::{Path, PathBuf};
use std::process::Command;
use std::time::{SystemTime, UNIX_EPOCH};

use crate::domain::RepoScan;

const README_CANDIDATES: &[&str] = &[
    "README.md",
    "README.txt",
    "README.rst",
    "readme.md",
    "readme.txt",
    "readme.rst",
];

const IGNORED_DIRS: &[&str] = &[
    ".git",
    ".deepwiki-mcp",
    "node_modules",
    "target",
    "__pycache__",
    ".next",
    "dist",
    "build",
];

pub fn scan_local_repository(repo_path: &Path) -> io::Result<RepoScan> {
    let repo_path = repo_path.canonicalize()?;
    let file_tree = collect_file_tree(&repo_path)?;
    let (readme_path, readme_content) = load_readme(&repo_path)?;
    let default_branch = detect_default_branch(&repo_path);

    Ok(RepoScan {
        scanned_at_epoch_ms: current_epoch_ms()?,
        root_path: repo_path,
        default_branch,
        readme_path,
        readme_content,
        file_tree,
    })
}

fn collect_file_tree(root: &Path) -> io::Result<Vec<String>> {
    let mut files = Vec::new();
    walk_dir(root, root, &mut files)?;
    files.sort();
    Ok(files)
}

fn walk_dir(root: &Path, dir: &Path, files: &mut Vec<String>) -> io::Result<()> {
    for entry in fs::read_dir(dir)? {
        let entry = entry?;
        let path = entry.path();
        let file_type = entry.file_type()?;

        if file_type.is_dir() {
            let dir_name = entry.file_name().to_string_lossy().to_string();
            if IGNORED_DIRS.iter().any(|ignored| *ignored == dir_name) {
                continue;
            }
            walk_dir(root, &path, files)?;
            continue;
        }

        if file_type.is_file() {
            let relative = path
                .strip_prefix(root)
                .unwrap_or(path.as_path())
                .to_string_lossy()
                .to_string();
            files.push(relative);
        }
    }
    Ok(())
}

fn load_readme(root: &Path) -> io::Result<(Option<PathBuf>, Option<String>)> {
    for candidate in README_CANDIDATES {
        let path = root.join(candidate);
        if path.exists() {
            let content = fs::read_to_string(&path)?;
            return Ok((Some(path), Some(content)));
        }
    }
    Ok((None, None))
}

fn detect_default_branch(root: &Path) -> Option<String> {
    let output = Command::new("git")
        .arg("-C")
        .arg(root)
        .arg("rev-parse")
        .arg("--abbrev-ref")
        .arg("HEAD")
        .output()
        .ok()?;

    if !output.status.success() {
        return None;
    }

    let branch = String::from_utf8_lossy(&output.stdout).trim().to_string();
    if branch.is_empty() {
        None
    } else {
        Some(branch)
    }
}

fn current_epoch_ms() -> io::Result<u128> {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|duration| duration.as_millis())
        .map_err(io::Error::other)
}

File: python/src/deepwiki_mcp/repo.py
"""Git operations and local filesystem scanning.

No LLM calls — the server only walks the tree and shells out to git so the
host model has a concrete inventory to work from.
"""
from __future__ import annotations

import os
import subprocess
import time
from pathlib import Path

from .models import RepoScan

README_CANDIDATES = (
    "README.md",
    "README.txt",
    "README.rst",
    "readme.md",
    "readme.txt",
    "readme.rst",
)

IGNORED_DIRS = frozenset(
    {
        ".git",
        ".deepwiki-mcp",
        "node_modules",
        "target",
        "__pycache__",
        ".next",
        "dist",
        "build",
        ".venv",
        "venv",
        ".mypy_cache",
        ".ruff_cache",
        ".pytest_cache",
    }
)


def git_clone(url: str, target_dir: Path) -> None:
    target_dir.parent.mkdir(parents=True, exist_ok=True)
    result = subprocess.run(
        ["git", "clone", "--depth=1", "--single-branch", url, str(target_dir)],
        check=False,
    )
    if result.returncode != 0:
        raise RuntimeError(f"git clone failed for {url}")


def git_pull(repo_dir: Path) -> None:
    result = subprocess.run(
        ["git", "-C", str(repo_dir), "pull", "--ff-only"],
        check=False,
    )
    if result.returncode != 0:
        raise RuntimeError(f"git pull failed for {repo_dir}")


def detect_default_branch(repo_dir: Path) -> str | None:
    try:
        result = subprocess.run(
            ["git", "-C", str(repo_dir), "rev-parse", "--abbrev-ref", "HEAD"],
            check=False,
            capture_output=True,
            text=True,
        )
    except FileNotFoundError:
        return None
    if result.returncode != 0:
        return None
    branch = result.stdout.strip()
    return branch or None


def _load_readme(root: Path) -> tuple[Path | None, str | None]:
    for candidate in README_CANDIDATES:
        path = root / candidate
        if path.exists():
            try:
                return path, path.read_text(encoding="utf-8", errors="replace")
            except OSError:
                return path, None
    return None, None


def _collect_file_tree(root: Path) -> list[str]:
    files: list[str] = []
    for dirpath, dirnames, filenames in os.walk(root):
        dirnames[:] = [d for d in dirnames if d not in IGNORED_DIRS]
        current = Path(dirpath)
        for name in filenames:
            rel = (current / name).relative_to(root).as_posix()
            files.append(rel)
    files.sort()
    return files


def scan_local_repository(repo_path: Path) -> RepoScan:
    resolved = repo_path.resolve()
    if not resolved.exists():
        raise FileNotFoundError(f"repository not found: {resolved}")
    file_tree = _collect_file_tree(resolved)
    readme_path, readme_content = _load_readme(resolved)
    default_branch = detect_default_branch(resolved)
    return RepoScan(
        scanned_at_epoch_ms=int(time.time() * 1000),
        root_path=resolved,
        default_branch=default_branch,
        readme_path=readme_path,
        readme_content=readme_content,
        file_tree=file_tree,
    )
